hie-plugin-api/Haskell/Ide/Engine/Ghc.hs

```diff
@@ -45,10 +47,23 @@ import           GHC
 import           IOEnv                             as G
 import           HscTypes
 import           Outputable                        (renderWithStyle)
+import           Language.Haskell.LSP.Types        ( NormalizedUri(..), toNormalizedUri )
 
 -- ---------------------------------------------------------------------
 
-type Diagnostics = Map.Map Uri (Set.Set Diagnostic)
+newtype Diagnostics = Diagnostics (Map.Map NormalizedUri (Set.Set Diagnostic))
+  deriving (Show, Eq)
+
+instance Semigroup Diagnostics where
+  Diagnostics d1 <> Diagnostics d2 = Diagnostics (d1 <> d2)
+
+instance Monoid Diagnostics where
+  mempty = Diagnostics mempty
```

Comment:
Can these be derived?

Reply:
The default monoid instance for Map is probably not what we want. This should be Map.Union with Set.union. the default semi group operation is leftbiased union.

Reply:
In any case I think `ToJSON` should be derivable. Although it looks like `NormalizedUri` has no instance for `ToJSON/FromJSON`. This should probably be added in haskell-lsp.